### ISSUE TITLE
Pump and curve types

### ISSUE THREAD

**Issue by eladsal** [OpenWaterAnalytics/EPANET#208]
Following the [question](http://community.wateranalytics.org/t/how-to-get-epanet-curve-type/680) on the forum, I have checked the header files and it seems that we have things wrong. The type [here](https://github.com/OpenWaterAnalytics/EPANET/blob/dev/include/epanet2.h#L256) are for pumps and not curves and we are missing the curve types in the file (they should be like [this](https://github.com/OpenWaterAnalytics/EPANET/blob/dev/src/types.h#L150)).

It is an easy fix but I'd like someone to confirm I'm not making a mess.

**Comment by LRossman** [comment]
@eladsal when you say "we are missing the curve types in the file" you mean the API header file, _epanet2.h_, right? If so, then the current `EN_CurveType` enum in _epanet2.h_ should be re-named to `EN_PumpType` and a new enum named `EN_CurveType` added for curve types. It would contain the same entries as `CurveType` in _types.h_ but with an _EN__ prefix. It should also contain an additional entry, `EN_G_CURVE` for generic curves that have not been assigned a type.

The above only needs to be done if one wants to add an `EN_getcurvetype` function in the API. Adding such a function would simply return the `Type` data field for a given Curve object. However it seems there is no way to specify what a curve's type should be either through reading the input file or through the `EN_addcurve` function. The `Type` field is undefined when a curve is created by reading an input file and is set to -1 when using the `EN_addcurve` function. So it seems adding a complementary `EN_setcurvetype` function would also be needed.

**Comment by milad-ghiami** [comment]
hi @LRossman   ,how are you?
im speak by @eladsal  about this tapic  in gmail.
you attention to enums on epanet and im think some chenge must to submit to epanet,
im seand for you   all emails   which have talk about this problem by @eladsal

**Comment by eladsal** [comment]
@LRossman, thank you, this is what I meant about the header file. However, what is a "generic curve", where is it used? The GPV uses a head curve, right?

Regarding the second part of your comment, when I added the `EN_setflowunits` function I had to know the types of the curves in order to convert their values to the new units. What I came up with is setting types to the curves when they are assigned to the different elements. For example, `V_CURVE` is [set](https://github.com/OpenWaterAnalytics/EPANET/blob/dev/src/input3.c#L205) in the `tankdata` function, `E_CURVE` is [set](https://github.com/OpenWaterAnalytics/EPANET/blob/dev/src/input3.c#L1446) in the `energydata` function and so on. This is not a perfect solution but covers the normal use of curves. I see that the `EN_addcurve` needs a fix when copying the curves list and also the `EN_setheadcurveindex` function should be fixed to set the curve type to `H_CURVE`. I would appreciate any comment you may have of this implementation.

Long story short, I think that @milad-ghiami is right about the need for the `EN_getcurvetype` function.

**Comment by LRossman** [comment]
I missed the fact that the v. 2.1 code was setting the curve type when a curve was assigned to an element. After looking at the code more closely here is a list of items that I think need fixing:
- `EN_addcurve` currently appends data to an existing curve. It should actually be able to create an entirely new curve of a specified type.` EN_setcurve` can be used to replace a curve's current data points.
- New functions `EN_getcurvetype` and `EN_setcurvetype` should be added to the API with `EN_CurveType` renamed to `EN_PumpType` in _epanet2.h_ and a new enumeration named `EN_CurveType` added with members _EN_V_CURVE, EN_P_CURVE_, etc.
- _EN_G_CURVE_ should be added to _epanet2.h_ and _G_CURVE_ to _types.h_ to represent a curve whose specific type is unknown, such as a curve that appears in the input file but is not used by any element.
- An error condition should be declared if an element tries to use a curve whose type is not applicable to the element.
- Checks should be added to functions like` EN_setflowunits` and `EN_addXXXX `that prevents them from executing when a hydraulic/water quality analysis is currently under way. This will require adding a new flag variable (like `RunFlag`) to the `EN_Project` structure that indicates if a simulation is currently running or not.

**Comment by eladsal** [comment]
@LRossman [`EN_addcurve`](https://github.com/OpenWaterAnalytics/EPANET/blob/dev/src/epanet.c#L2886) actually creates a new curve with one data point of (1,1).

**Comment by eladsal** [comment]
Some comments to @LRossman suggestions:

1. `EN_addcurve` actually creates a new curve with one data point of (1,1). However, curve type should be handled when copying the curves list.
2. Apart from the `EN_setcurvetype` function I agree. If the curve type isn't saved to the INP file and is set according to its use, what would be the use case for the user to set the type? Maybe for other programs to use. If we add it I'm not sure that there is would be right to raise an error when misused (see item 4).
3. So the general curve will be the new default instead of -1.
4. Data consistency are currently done only when the INP file is read, right? If we add the `EN_setcurvetype` function, the checks should be in it (check where a curve is used). Users might be using one curve for different type in some "magic tricks", an error would prevent that use. Maybe a warning?
5. Where would such a flag be set, is it in `EN_initH` and unset in `EN_closeH`?

Thanks!

**Comment by milad-ghiami** [comment]
i try to get last version to check this, i use by version of 3 mone ago

**Comment by LRossman** [comment]
Regarding @eladsal comments:

1. Sorry, I mis-read the `EN_addcurve` code (blame it on old age) - of course it creates a new curve object .
2. I believe an `EN_setcurvetype` function is still needed for curves created with `EN_addcurve`, unless the curve type is added as an argument to `EN_addcurve` (which is what I would recommend).
3. Yes, general curve would be the new default.
4. Yes, consistent use of curves would require that an error condition be raised at the first instance where a wrong type of curve was used. However, to maintain backwards compatibility I suggest that this feature not be implemented at this time.
5. The `RunFlag` would be set in `EN_openH, EN_closeH, EN_openQ` and `EN_closeQ`. It would require careful thought to identify which API functions, especially those that modify parameters, would be affected by it, since for some parameters it makes sense to allow a change during an extended period analysis but for others it doesn't.

For @milad-ghiami , please read the [main page](https://github.com/OpenWaterAnalytics/EPANET) of this project to learn how to contribute to it.

**Comment by milad-ghiami** [comment]
In the second part, Mr. @eladsal talked as correctly  about my problem ,
i dont know curve type when load inp file  by API functions,
must to add new tag to inp file to show curve type

**Comment by LRossman** [comment]
@milad-ghiami please look at the fourth comment in this thread. @eladsal infers what a curve's type is during the input file loading process by seeing which type of element refers to it. Remember, the only place in the code where curve type is used right now is in the `EN_setflowunits` function, where it determines units conversion factors for each curve's data points. So I don't think it is worth changing the format of the input file just for this (plus it would cause problems for the EPANET 2.0 GUI).

**Comment by eladsal** [comment]
@milad-ghiami if you are getting -1 for all curves (try Net1.inp) it means that you are probably not using the latest `dev` branch of the EPANET repository.

**Comment by milad-ghiami** [comment]
@eladsal yes i use epanet version  20100
 i will get last version and check agian

**Comment by eladsal** [comment]
If there are no additional comments I will go ahead and apply the changes and commit. This will include:
1. adding a new `G_CURVE` curve type to `types.h` for curves with no assigned type (replaces the -1 value).
2. `EN_G_CURVE` will be added to `epanet2.h`.
3. `EN_CurveType` will be renamed to `EN_PumpType` in `epanet2.h` and a new enumeration named `EN_CurveType` will be added with members `EN_V_CURVE`, `EN_P_CURVE`, etc.

**Comment by milad-ghiami** [comment]
i think you must define curve type when you add new curve by addcurve function   
 and not  permit to  it to use for another type

**Comment by eladsal** [comment]
@milad-ghiami, the `ENaddcurve` is already in release 2.1. Adding another variable to the function call would break backward compatibility, a thing I'm not sure we want to do. The workaround would be to add the `ENsetcurvetype` function and call it.

**Comment by milad-ghiami** [comment]
@eladsal   getcurvetype and will  add to API
